--- libzvt-2.0.0/libzvt/Makefile.am.wiget	Thu Jun 20 10:18:04 2002
+++ libzvt-2.0.0/libzvt/Makefile.am	Thu Jun 20 10:19:33 2002
@@ -74,10 +74,9 @@
 
 gnome_pty_helper_SOURCES=   \
 	gnome-pty-helper.c    \
-        gnome-login-support.c \
         gnome-utmp.c
 
-gnome_pty_helper_LDADD= $(UTIL_LIBS)
+gnome_pty_helper_LDADD= $(UTIL_LIBS) libzvt-2.0.la
 
 # Since we do gnome-pty-helper manually, include the relevant sources here
 EXTRA_DIST = TODO BUGS README	\
